File: iso27001-phoenix/test/unit/encryption_test.exs
defmodule Iso27001Phoenix.EncryptionTest do
  use ExUnit.Case, async: true

  alias Iso27001Phoenix.Infrastructure.Encryption.FieldEncryptor

  # Set 32-byte key in application env before tests
  setup do
    Application.put_env(:iso27001_phoenix, :encryption_key, "test-only-key-exactly-32-bytes!!")
    :ok
  end

  test "round-trip encrypt/decrypt returns original plaintext" do
    plaintext  = "user@example.com"
    ciphertext = FieldEncryptor.encrypt(plaintext)
    assert {:ok, ^plaintext} = FieldEncryptor.decrypt(ciphertext)
  end

  test "each encryption produces a different ciphertext (unique IV)" do
    plaintext = "same-input"
    c1 = FieldEncryptor.encrypt(plaintext)
    c2 = FieldEncryptor.encrypt(plaintext)
    refute c1 == c2
  end

  test "wrong key length raises at encrypt time" do
    Application.put_env(:iso27001_phoenix, :encryption_key, "too-short")
    assert_raise RuntimeError, ~r/32 bytes/, fn ->
      FieldEncryptor.encrypt("anything")
    end
  after
    Application.put_env(:iso27001_phoenix, :encryption_key, "test-only-key-exactly-32-bytes!!")
  end

  test "tampered ciphertext returns decryption error" do
    ciphertext = FieldEncryptor.encrypt("hello")
    tampered   = ciphertext <> "X"
    assert {:error, :decryption_failed} = FieldEncryptor.decrypt(tampered)
  end
end

File: iso27001-gin/tests/unit/encryption_test.go
package unit_test

import (
	"testing"

	"github.com/iso27001/gin-blueprint/internal/infrastructure/encryption"
)

func TestFieldEncryptor_RoundTrip(t *testing.T) {
	enc, err := encryption.New("test-only-key-exactly-32bytes-ab")
	if err != nil {
		t.Fatalf("New() failed: %v", err)
	}

	plaintext := "user@example.com"
	ciphertext, err := enc.Encrypt(plaintext)
	if err != nil {
		t.Fatalf("Encrypt() failed: %v", err)
	}

	if ciphertext == plaintext {
		t.Error("ciphertext should differ from plaintext")
	}

	decrypted, err := enc.Decrypt(ciphertext)
	if err != nil {
		t.Fatalf("Decrypt() failed: %v", err)
	}

	if decrypted != plaintext {
		t.Errorf("expected %q, got %q", plaintext, decrypted)
	}
}

func TestFieldEncryptor_DifferentIVEachEncryption(t *testing.T) {
	enc, _ := encryption.New("test-only-key-exactly-32bytes-ab")
	plaintext := "same-value"

	ct1, _ := enc.Encrypt(plaintext)
	ct2, _ := enc.Encrypt(plaintext)

	if ct1 == ct2 {
		t.Error("each encryption should produce a different ciphertext (random IV)")
	}
}

func TestFieldEncryptor_InvalidKeyLength(t *testing.T) {
	_, err := encryption.New("short-key")
	if err == nil {
		t.Error("expected error for key != 32 bytes")
	}
}
